### ISSUE TITLE
java.lang.IllegalStateException: Cannot use ImplicitTransaction after it or its parent has been closed.

### ISSUE THREAD

**Issue by ImGauravC** [realm/realm-java#726]
```java
java.lang.IllegalStateException: Cannot use ImplicitTransaction after it or its parent has been closed.
            at io.realm.internal.ImplicitTransaction.assertNotClosed(ImplicitTransaction.java:66)
            at io.realm.internal.ImplicitTransaction.advanceRead(ImplicitTransaction.java:29)
            at io.realm.Realm$RealmCallback.handleMessage(Realm.java:194)
            at android.os.Handler.dispatchMessage(Handler.java:95)
            at android.os.Looper.loop(Looper.java:176)
```

I'm getting the above error while adding/updating realm objects in a background handler thread. What is the reason behind this?

**Comment by cmelchior** [comment]
You have called close() on Realm and are probably trying to access a RealmResult or RealmObject afterwards.

**Comment by ImGauravC** [comment]
Is this the only possibility? Because I don't see this happening anywhere in my code.

**Comment by cmelchior** [comment]
It is by far the most likely. We will need more information to be more specific, can you share your code or try to reduce it to a testcase?

**Comment by ImGauravC** [comment]
I got it finally. Thanks :)

**Comment by ImGauravC** [comment]
Wait, encountering it again. Below is an explanation of what I'm trying to do. 

I need to constantly update the app with data from the server. So for polling data, I created a handler thread with a Looper.  On app start, in the run() method of the Runnable, I get a Realm instance, perform a transaction, close the Realm instance, and then post the Runnable again to the same thread with a delay of say, 60 seconds. 

Below is some pseudo code:

```java
Runnable r = new Runnable() {
     @Override
     public void run() {
         httpRequest.get(uri, new DataCallback() {
             public void onSuccess(JSONObject result) throws JSONException {
                 Realm realm = Realm.getInstance(context);
                 realm.setAutoRefresh(true);
                 try {
                     realm.beginTransaction();
                     <Some write with data from the server>
                     realm.commitTransation();
                 } catch (Exception e) {
                     realm.cancelTransaction();
                 } 
                 realm.close();
             }
        }
        handlerThread.postDelayed(r, 60 * 1000);
    } 
}  

handlerThread.post(r)
```

**Comment by cmelchior** [comment]
That pattern looks pretty solid. You just have to be careful about catching Exception, as you might catch OutOfMemory which could potentially cause corruption on your database. Catching RuntimeException is much safer.

Is it inside that exact runnable you get the ImplicitTransaction exception?

**Comment by cmelchior** [comment]
@ImGauravC 
What kind of writes are doing. Only writes on new objects or overriding objects that came from somewhere else?

**Comment by ImGauravC** [comment]
@cmelchior : I sync current state with the server response. So it involves either:
a) using setters to change values of RealmObjects.
b) creating new Realm Objects
c) Since some data might be deleted on the server, so in order to keep the data in sync on the client, I delete all realm objects using realm.clear(<ObjectClass>) and then recreate them with the server response.

**Comment by ImGauravC** [comment]
Oh, and the objects might be in use in activity/adapter while this is being done.

**Comment by cmelchior** [comment]
You a) might be causing the problems. Realm instances are cached on the same thread and reference counted, but if you reference count reaches 0, it will close the Realm and all objects are no longer valid.

So if you access RealmObjects you have not queried for in the success handler you might risk they have been closed.

**Comment by ImGauravC** [comment]
@cmelchior : I do query for the objects in the success handler. Still facing the issue.

**Comment by cmelchior** [comment]
In that case I would be very interested to know exactly what your code does as you might have stumbled upon a bug somewhere. Are you in a position to share it or can you reduce it to a minimal case?

Just a note. I just found a bug in our our Realm.close() code that could cause it to behave wrong if you call close to many times, so I would check for that.

**Comment by cmelchior** [comment]
@ImGauravC 
I found another bug in how our Realm.close() method works on background threads which could explain the behaviour you are seeing. See the above PR for a fix.

**Comment by ImGauravC** [comment]
Great! Can we have the 0.77 soon?

**Comment by bmunkholm** [comment]
Sure - we hope to cut it tomorrow all going well :-)

**Comment by cmelchior** [comment]
Fix has been merged into master and will be part of 0.77.0

**Comment by ImGauravC** [comment]
Hi @cmelchior 
I'm still seeing this.
```java
java.lang.IllegalStateException: Cannot use ImplicitTransaction after it or its parent has been closed.
       at io.realm.internal.ImplicitTransaction.assertNotClosed(ImplicitTransaction.java:66)
       at io.realm.internal.ImplicitTransaction.advanceRead(ImplicitTransaction.java:29)
       at io.realm.Realm$RealmCallback.handleMessage(Realm.java:212)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:137)

**Comment by cmelchior** [comment]
Also with the newest 0.78 release?

**Comment by ImGauravC** [comment]
Didn't check this one

**Comment by ImGauravC** [comment]
Yes, even with the new version.

**Comment by cmelchior** [comment]
Hi @ImGauravC 
Are you still seeing this issue with the latest version of Realm?

**Comment by cmelchior** [comment]
Closing due to no response. Please reopen if there is any new information.

**Comment by andrewthecoder** [comment]
Hi @cmelchior, not sure whether I'm experiencing the same issue again or if I'm just doing something wrong, but I'm finding it very hard to debug and I'm not sure what to look at next.

Here's the stack trace I'm getting from my app (after recently rewriting/refactoring it to use Realm instead of SQLite, to avoid limitations and bug we were facing with SQLite):

```
java.lang.IllegalStateException: Cannot use ImplicitTransaction after it or its parent has been closed.
at io.realm.internal.ImplicitTransaction.assertNotClosed(ImplicitTransaction.java:66)
at io.realm.internal.ImplicitTransaction.advanceRead(ImplicitTransaction.java:29)
at io.realm.Realm$RealmCallback.handleMessage(Realm.java:283)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:137)
at android.os.HandlerThread.run(HandlerThread.java:60)
```

From that stack trace, how should I go about figuring out what part of my code is accessing a RealmObject after closing the Realm?
I'm fairly confident I'm not, as all of my reads/writes of data use the following pattern:

```
Realm realm = Realm.getDefaultInstance();
try {}
catch (Exception e)  {}
finally
{
     if (realm != null)
     {
            realm.close();
     }
}
```

Unfortunately I can't easily reproduce the issue, but it's happened in various situations while testing our app. 
I'd be happy to share my code with you privately if it would help; please email me if this is the case.

**Comment by andrewthecoder** [comment]
I forgot to mention - I'm using Realm 0.82.1 from Maven.
Thanks in advance for any suggestions - I'm under quite a bit of pressure to try and get this release of our application ready for deployment within a few days!

**Comment by cmelchior** [comment]
Hi @andrewthecoder 
Note that as Realm is a zero-copy database. We require a open connection to the database as long as you want to access any data. That means the above pattern is a bit dangerous as it is very easy to save references to Realm objects outside the scope of the `try/catch`. Instead I would recommend you do put `Realm.getInstance` in `onCreate` and `realm.close()` in `onDestroy` in all your activities. That will ensure that you Realm is open for the duration of the app and closed properly when exiting.

If that doesn't work, feel free to share the code privately with help@realm.io where it might be easier to help.

**Comment by andrewthecoder** [comment]
Thanks for the reply @cmelchior, I understand the concept but wasn't sure of a better way to solve it, as our application has multiple background threads as well as activities.

The onCreate/onDestroy solution makes sense to me for activities, but I wasn't sure how to properly handle simultaneous access to the database from the background threads.

I'll try and refactor the app today to open a single connection to the database for each activity and service.
My worry was that as I'd read Realm wasn't thread-safe, accessing the database with one connection from the UI thread and another from a background thread would cause issues?
If that's not the case then great, I guess I just need to refactor!

In general though, how would you go about figuring out where you're accessing a realm object incorrectly from the stack trace I posted above, or is this a limitation of the message handling design?

**Comment by cmelchior** [comment]
Realm is an MVCC database which allow concurrent access. So it shouldn't be a problem using one database connection for the background thread and one for the UI thread. It is actually the right approach. Realm will automatically detect if you try to access data from a wrong thread and throw an exception in that case.

Regarding your stack trace above. It looks like you have registered a RealmChangeListener that somehow gets triggered even after the realm instance on that thread has been closed. That is something we will have to look into as that might be a bug on our side. To be safe you can unregister any listener when you close the Realm instance.

**Comment by andrewthecoder** [comment]
Thanks for your advice @cmelchior, I've refactored our app's usage of realm to be getting a realm instance and closing it in a try/catch/finally at a very low level, which has solved almost all of the issues I was having with it as we're no longer creating realm instances which aren't being closed!

However, unfortunately today one of our test devices crashed with the same error/stack trace as I posted above, as in the title of this issue, leading me to believe this issue is not yet solved.
```
java.lang.IllegalStateException: Cannot use ImplicitTransaction after it or its parent has been closed.
at io.realm.internal.ImplicitTransaction.assertNotClosed(ImplicitTransaction.java:66)
at io.realm.internal.ImplicitTransaction.advanceRead(ImplicitTransaction.java:29)
at io.realm.Realm$RealmCallback.handleMessage(Realm.java:283)
at android.os.Handler.dispatchMessage(Handler.java:98)
at android.os.Looper.loop(Looper.java:146)
at android.os.HandlerThread.run(HandlerThread.java:61)
```

Here's the surrounding log lines from my crash report, in case it's helpful. 
```
1947	08-26 13:27:33.199 D/PF.HttpRequestAsyncTask(16099): Adding HTTP parameter name value pair; name = iat, value = 1
1948	08-26 13:27:33.199 D/PF.HttpRequestAsyncTask(16099): Adding HTTP parameter name value pair; name = v, value = 1.3
1949	08-26 13:27:33.199 D/PF.HttpRequestAsyncTask(16099): Adding HTTP parameter name value pair; name = d, value = 11423
1950	08-26 13:27:34.019 D/dalvikvm(16099): GC_FOR_ALLOC freed 1647K, 74% free 10849K/40560K, paused 55ms, total 55ms
1951	08-26 13:27:34.019 I/System.out(16099): AsyncTask #2 calls detatch()
1952	08-26 13:27:34.024 D/PF.HttpRequestAsyncTask(16099): onPostExecute, returnStatus = true
1953	08-26 13:27:34.024 D/PF.JavaScriptInterface(16099): GetJobsData onTaskCompleted returnStatus: true
1954	08-26 13:27:34.024 D/PF.JavaScriptInterface(16099): returnStatus was true, checking responseString is not empty
1955	08-26 13:27:34.069 I/PF.JavaScriptInterface(16099): Executing JavaScript in WebView: PFCallbacks.listJobsSuccess() from com.podfather.podfather.interfaces.JavaScriptInterface.callJSInWebview, com.podfather.podfather.interfaces.JavaScriptInterface.onTaskCompleted, com.podfather.podfather.tasks.HttpRequestAsyncTask.onPostExecute
1956	08-26 13:27:34.074 D/PF.JavaScriptInterface(16099): JS: listJobsSuccess: nativeWrapperlistJobsSuccess called, getting jobsData object from DB
1957	08-26 13:27:34.079 D/PF.DataPostService(16099): starting dataPostService as no other data post is running
1958	08-26 13:27:34.079 D/PF.DataPostService(16099): onHandleIntent setting dataPostService.currentlyProcessing to true
1959	08-26 13:27:34.079 D/PF.DataPostService(16099): Network is available, about to post any data waiting to be sent
1960	08-26 13:27:34.139 D/PF.JavaScriptInterface(16099): In method: removeStorageString, removing key: listJobsJSON from localStorage
1961	08-26 13:27:34.434 D/PF.JavaScriptInterface(16099): JS: listJobsSuccess: newJobTemplateObjects contains at least one new object, iterating through
1962	08-26 13:27:34.439 D/PF.JavaScriptInterface(16099): JS: listJobsSuccess: listJobs contained 0 new jobTemplates, and 5 which we already have stored. No change to jobTemplates object.
1963	08-26 13:27:34.444 I/PF.JavaScriptInterface(16099): JS: listJobsSuccess: No newRunObjects found
1964	08-26 13:27:34.449 I/PF.JavaScriptInterface(16099): JS: listJobsSuccess: No newJobObjects found
1965	08-26 13:27:34.454 D/PF.JavaScriptInterface(16099): JS: listJobsSuccess: newTasks is iterable, looking for implemented task types
1966	08-26 13:27:34.454 D/PF.JavaScriptInterface(16099): JS: listJobsSuccess: No pending_job_deletions tasks found
1967	08-26 13:27:34.499 D/PF.DataPostService(16099): Successful POST requests: 0, failed POST requests: 0
1968	08-26 13:27:34.499 D/PF.DataPostService(16099): calling JS function: PFCallbacks.actionDataPostComplete()
1969	08-26 13:27:34.529 I/PF.JavaScriptInterface(16099): Executing JavaScript in WebView: PFCallbacks.actionDataPostComplete(0, 0) from com.podfather.podfather.interfaces.JavaScriptInterface.callJSInWebview, com.podfather.podfather.services.DataPostService.onHandleIntent, android.app.IntentService$ServiceHandler.handleMessage
1970	08-26 13:27:34.529 D/PF.DataPostService(16099): onHandleIntent setting dataPostService.currentlyProcessing to false
1971	08-26 13:27:34.534 D/PF.DataPostService(16099): Completed dataPostService, releasing wakelock if one is held
1972	08-26 13:27:34.534 W/dalvikvm(16099): threadid=33: thread exiting with uncaught exception (group=0x41a62c08)
1973	08-26 13:27:34.539 E/ACRA (16099): ACRA caught a IllegalStateException for com.podfather.podfather
1974	08-26 13:27:34.539 E/ACRA (16099): java.lang.IllegalStateException: Cannot use ImplicitTransaction after it or its parent has been closed.
1975	08-26 13:27:34.539 E/ACRA (16099): at io.realm.internal.ImplicitTransaction.assertNotClosed(ImplicitTransaction.java:66)
1976	08-26 13:27:34.539 E/ACRA (16099): at io.realm.internal.ImplicitTransaction.advanceRead(ImplicitTransaction.java:29)
1977	08-26 13:27:34.539 E/ACRA (16099): at io.realm.Realm$RealmCallback.handleMessage(Realm.java:283)
1978	08-26 13:27:34.539 E/ACRA (16099): at android.os.Handler.dispatchMessage(Handler.java:98)
1979	08-26 13:27:34.539 E/ACRA (16099): at android.os.Looper.loop(Looper.java:146)
1980	08-26 13:27:34.539 E/ACRA (16099): at android.os.HandlerThread.run(HandlerThread.java:61)
1981	08-26 13:27:34.539 D/ACRA (16099): Building report
1982	08-26 13:27:34.549 D/ACRA (16099): Using custom Report Fields
1983	08-26 13:27:34.574 D/PF.JavaScriptInterface(16099): JS: actionDataPostComplete: Data post complete and user is logged in, updating datatosend and triggering listJobs if necessary
1984	08-26 13:27:34.599 D/Toast (16099): checkMirrorLinkEnabled returns : false
1985	08-26 13:27:34.599 D/Toast (16099): showing allowed
1986	08-26 13:27:34.634 D/dalvikvm(16099): GC_CONCURRENT freed 2219K, 74% free 10609K/40560K, paused 6ms+7ms, total 55ms
1987	08-26 13:27:34.674 D/PF.JavaScriptInterface(16099): dataRowsCount: 0
1988	08-26 13:27:34.784 D/PF.JavaScriptInterface(16099): JS: communicationDue: communicationDue called, returning false as lastCommunicationTimestamp was only 0 seconds (less than 2 mins) ago
1989	08-26 13:27:34.964 W/Settings(16099): Setting call_auto_retry has moved from android.provider.Settings.System to android.provider.Settings.Global, returning read-only value.
1990	08-26 13:27:35.354 W/Settings(16099): Setting emergency_tone has moved from android.provider.Settings.System to android.provider.Settings.Global, returning read-only value.
1991	08-26 13:27:35.469 W/Settings(16099): Setting font_size has moved from android.provider.Settings.System to android.provider.Settings.Global, returning read-only value.
1992	08-26 13:27:36.769 W/Settings(16099): Setting roaming_settings has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
1993	08-26 13:27:36.794 W/Settings(16099): Setting data_roaming2 has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
1994	08-26 13:27:36.844 W/Settings(16099): Setting mobile_data has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
1995	08-26 13:27:36.844 W/Settings(16099): Setting mobile_data_secondary has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
1996	08-26 13:27:36.889 W/Settings(16099): Setting setup_prepaid_data_service_url has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
1997	08-26 13:27:36.924 W/Settings(16099): Setting wifi_saved_state has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
1998	08-26 13:27:37.129 I/ACRA (16099): READ_LOGS granted! ACRA can include LogCat and DropBox data.
1999	08-26 13:27:37.234 D/ACRA (16099): Retrieving logcat output...
```

I'm now very confident that our app isn't misusing Realm or doing anything weird, and all instances are being closed correctly. 
See these two screenshots from my IDE:
![Instance creation](http://i.imgur.com/0HC78Yo.png) ![Instance closure](http://i.imgur.com/rEy9SMf.png)

I'm definitely closing the realm instance after every usage, all of those 11 usages are very similar to this example:
```
Realm realm = Realm.getDefaultInstance();

try
{
    LocalStorageItem localStorageItem = realm.where(LocalStorageItem.class).equalTo("key", key).findFirst();

    if (localStorageItem != null)
    {
        value = localStorageItem.getValue();
    }
}
catch(Exception e)
{
    CustomLog.exception(LOG_TAG, e);
}
finally
{
    realm.close();
}
```


I'd appreciate any help you can provide, and if you think me sharing our full source code with you by email would help solve this, I can try and arrange something.

The biggest frustration is there not being any useful information in the stack trace - how would you even go about trying to solve this, compile Realm myself and try and debug it?

**Comment by dzeikei** [comment]
I'm having the same exception with Realm running in background thread as part of an IntentService. It occurs intermittently and the app may have realm transactions happening in other threads at the same time.

    @Override
    protected void onHandleIntent(Intent intent) {
        if (intent.getAction() != null) {
            Realm realm = null;
            try {
                realm = Realm.getDefaultInstance();
                // Get realm objects here and convert to JSON using GSON and send to server
            } finally {
                if (realm != null) {
                    realm.close();
                }
            }
        }
    }

**Comment by andrewthecoder** [comment]
Can we open this again as both myself and @dzeikei have experienced it recently?

Or, is this from the 0.83 changelog relevant? 
"Opening a Realm file from one thread will no longer be blocked by a write transaction from another thread."

**Comment by andrewthecoder** [comment]
Just to confirm this is still an issue with realm-android-0.83.0-20150912.231049

I'm about to start looking into either fixing this myself, or at least catching the exception silently - any suggestions as to where I should start or what the repercussions of allowing this to happen silently are?

**Comment by cmelchior** [comment]
Hi @andrewthecoder We will look into it. Catching the exception and ignoring should be fine, as the exception says the Realm has already been closed, so you will not be leaking resources.

**Comment by andrewthecoder** [comment]
Thanks for the confirmation @cmelchior - this seems to have "fixed" the issue for us at least. For anyone else in this thread looking for a quick and dirty solution for now, my build can be downloaded from here: http://crosswalk.beveridge.uk/realm-java/realm/build/libs/

**Comment by cmelchior** [comment]
Note that this is most likely triggered by the following code:

```
    private class RealmCallback implements Handler.Callback {
        @Override
        public boolean handleMessage(Message message) {
            if (message.what == REALM_CHANGED) {
                transaction.advanceRead();
                sendNotifications();
            }
            return true;
        }
    }
```

`transaction` will throw this error if the Realm is closed. I am still trying to figure out how it is happening though as your `close()` method should clean the Handler for any pending messages.

**Comment by cmelchior** [comment]
I found the error. We are keeping an internal cache of Handlers and if you open the same Realm on multiple Handler threads, only the last Handler will be saved. This is wrong as it will potentially result in messages being posted to the wrong handler which in turn will try to update a Realm that is closed.

@nhachicha This seems to be fixed by your PR #1214? Should we split that fix into a separate PR or just keep it as part of yours?

**Comment by cmelchior** [comment]
Ups, It was already split into a separate PR. A fix for this is here: https://github.com/realm/realm-java/issues/1467

**Comment by cmelchior** [comment]
@andrewthecoder We just merged a fix for this to master. So it should be available in our SNAPSHOT release within 15-30 minutes.
